lsp-mode.el

```diff
@@ -2028,6 +2030,242 @@ CALLBACK - callback for the lenses."
     (setq-local lsp--lens-last-count nil)
     (setq-local lsp--lens-backend-cache nil))))
 
+
+;; toggles
+
+(defun lsp-toggle-trace-io ()
+  "Toggle client-server protocol logging."
+  (interactive)
+  (setq lsp-print-io (not lsp-print-io))
+  (lsp--info "Server logging %s." (if lsp-print-io "enabled" "disabled")))
+
+(defun lsp-toggle-signature-auto-activate ()
+  "Toggle signature auto activate."
+  (interactive)
+  (setq lsp-signature-auto-activate (not lsp-signature-auto-activate))
+  (lsp--info "Signature autoactivate %s." (if lsp-signature-auto-activate "enabled" "disabled"))
+  (lsp--update-signature-help-hook))
+
+(defun lsp-toggle-on-type-formatting ()
+  "Toggle on type formatting."
+  (interactive)
+  (setq lsp-enable-on-type-formatting (not lsp-enable-on-type-formatting))
+  (lsp--info "On type formatting is %s." (if lsp-enable-on-type-formatting "enabled" "disabled"))
+  (lsp--update-on-type-formatting-hook))
+
+(defun lsp-toggle-symbol-highlight ()
+  "Toggle symbol highlighting."
+  (interactive)
+  (setq lsp-enable-symbol-highlighting (not lsp-enable-symbol-highlighting))
+
+  (cond
+   ((and lsp-enable-symbol-highlighting
+         (lsp-feature? "textDocument/documentHighlight"))
+    (add-hook 'lsp-on-idle-hook #'lsp--document-highlight nil t)
+    (lsp--info "Symbol highlighting enabled in current buffer."))
+   ((not lsp-enable-symbol-highlighting)
+    (remove-hook 'lsp-on-idle-hook #'lsp--document-highlight t)
+    (lsp--remove-overlays 'lsp-highlight)
+    (lsp--info "Symbol highlighting disabled in current buffer."))))
+
+
+;; keybindings
+
+(defmacro lsp-define-conditional-key (keymap key def
+                                             &rest body)
+  "In KEYMAP, define key sequence KEY as DEF conditionally.
+This is like `define-key', except the definition
+\"disappears\" whenever BODY evaluates to nil."
+  (declare (indent 3)
+           (debug (form form form &rest sexp)))
+
+  `(define-key ,keymap ,key
+     '(menu-item
+       ,(format "maybe-%s" (or (car (cdr-safe def)) def))
+       ,def
+       :filter (lambda (item)
+                 (when (with-current-buffer lsp--describe-buffer
+                         ,@body)
+                   item)))))
+
+(defvar lsp--describe-buffer nil)
+
+(defun lsp-describe-buffer-bindings-advice (buffer &optional _prefix _menus)
+  (setq lsp--describe-buffer buffer))
+
+(add-function :before
```

Comment:
nit: why use `add-function` when `advice-add` covers more cases?

Reply:
I guess the answer is that it covers the requirements for this case. Is there any particular benefit from using advice-add in this case?

Reply:
https://www.gnu.org/software/emacs/manual/html_node/elisp/Advising-Named-Functions.html
> they know how to deal with macros and autoloaded functions, they let describe-function preserve the original docstring as well as document the added advice, and they let you add and remove advice before a function is even defined.

Reply:
Done - I guess we may benefit if the function is not loaded.